athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneMetadataHandler.java

```diff
@@ -281,37 +282,69 @@ public GetTableResponse doGetQueryPassthroughSchema(BlockAllocator allocator, Ge
 
         GetTableResponse getTableResponse = doGetTable(allocator, request);
         List<Field> fields = getTableResponse.getSchema().getFields();
-        Client client = neptuneConnection.getNeptuneClientConnection();
-        GraphTraversalSource graphTraversalSource = neptuneConnection.getTraversalSource(client);
-        String gremlinQuery = qptArguments.get(NeptuneQueryPassthrough.QUERY);
-        gremlinQuery = gremlinQuery.concat(".limit(1)");
-        logger.info("NeptuneMetadataHandler doGetQueryPassthroughSchema gremlinQuery with limit: " + gremlinQuery);
-        Object object = new PropertyGraphHandler(neptuneConnection).getResponseFromGremlinQuery(graphTraversalSource, gremlinQuery);
-        GraphTraversal graphTraversalForSchema = (GraphTraversal) object;
-        Map graphTraversalObj = null;
         Schema schema;
-        while (graphTraversalForSchema.hasNext()) {
-            graphTraversalObj = (Map) graphTraversalForSchema.next();
+        Enums.GraphType graphType = Enums.GraphType.PROPERTYGRAPH;
+
+        if (configOptions.get(Constants.CFG_GRAPH_TYPE) != null) {
+            graphType = Enums.GraphType.valueOf(configOptions.get(Constants.CFG_GRAPH_TYPE).toUpperCase());
         }
 
-        //In case of gremlin query is fetching all columns then we can use same schema from glue
-        //otherwise we will build schema from gremlin query result column names.
-        if (graphTraversalObj != null && graphTraversalObj.size() == fields.size()) {
+        switch (graphType){
+            case PROPERTYGRAPH:
+                Client client = neptuneConnection.getNeptuneClientConnection();
+                GraphTraversalSource graphTraversalSource = neptuneConnection.getTraversalSource(client);
+                String gremlinQuery = qptArguments.get(NeptuneQueryPassthrough.QUERY);
+                gremlinQuery = gremlinQuery.concat(".limit(1)");
```

Comment:
Why do we not check if the query containts `limit` like we do for the sparqlQuery on line 310?

Reply:
Hi @aimethed, in gremlin even if we have limit in query, it will override previous limit with appended limit. Whereas in sparql it won't override limit so explicitly checking.